pkg/operator/handlers/dapr_handler.go

```diff
@@ -30,43 +34,129 @@ const (
 	clusterIPNone                   = "None"
 )
 
+const (
+	daprServiceOwnerKey = ".metadata.controller"
+)
+
 var log = logger.NewLogger("dapr.operator.handlers")
 
 // DaprHandler handles the lifetime for Dapr CRDs
 type DaprHandler struct {
-	kubeAPI         *kubernetes.API
-	deploymentsLock *sync.Mutex
+	mgr ctrl.Manager
+
+	client.Client
+	Scheme *runtime.Scheme
 }
 
 // NewDaprHandler returns a new Dapr handler
-func NewDaprHandler(kubeAPI *kubernetes.API) *DaprHandler {
+func NewDaprHandler(mgr ctrl.Manager) *DaprHandler {
 	return &DaprHandler{
-		kubeAPI:         kubeAPI,
-		deploymentsLock: &sync.Mutex{},
+		mgr: mgr,
+
+		Client: mgr.GetClient(),
+		Scheme: mgr.GetScheme(),
 	}
 }
 
 // Init allows for various startup tasks
 func (h *DaprHandler) Init() error {
-	return nil
+	if err := h.mgr.GetFieldIndexer().IndexField(
+		&corev1.Service{}, daprServiceOwnerKey, func(rawObj runtime.Object) []string {
+			svc := rawObj.(*corev1.Service)
+			owner := meta_v1.GetControllerOf(svc)
+			if owner == nil || owner.APIVersion != appsv1.SchemeGroupVersion.String() || owner.Kind != "Deployment" {
+				return nil
+			}
+			return []string{owner.Name}
+		}); err != nil {
+		return err
+	}
+
+	return ctrl.NewControllerManagedBy(h.mgr).
+		For(&appsv1.Deployment{}).
+		Owns(&corev1.Service{}).
+		Complete(h)
+}
+
+func (h *DaprHandler) daprServiceName(appID string) string {
+	return fmt.Sprintf("%s-dapr", appID)
+}
+
+// Reconcile the expected service for dapr deployment
+func (h *DaprHandler) Reconcile(req ctrl.Request) (ctrl.Result, error) {
+	ctx := context.Background()
+
+	var deployment appsv1.Deployment
+	var expectedService bool
+	if err := h.Get(ctx, req.NamespacedName, &deployment); err != nil {
+		if apierrors.IsNotFound(err) {
+			log.Debugf("deployment has be deleted, %s", req.NamespacedName)
+			expectedService = false
+		} else {
+			log.Errorf("unable to get deployment, %s, err: %s", req.NamespacedName, err)
+			return ctrl.Result{}, err
+		}
+	} else {
+		if deployment.DeletionTimestamp != nil {
+			log.Debugf("deployment is being deleted, %s", req.NamespacedName)
+			expectedService = false
```

Comment:
Should we wait until the deployment is deleted before starting to delete the Service?